### ISSUE TITLE
Various Small Fixes

### ISSUE THREAD

**Issue by c0nn3r** [kuangliu/pytorch-retinanet#10]
- New convolution layer's initializations doesn't match the original RetinaNet paper. The `bias` should be set to `0` and the weights to a normal distribution weight fill with `σ = 0.01`.
- The last layer of classification and object detection SubNets don't have proper initialization (this might explain the NaN errors, as they noted in the paper that not doing this causes them). The bias of each should be set to (according to the π value selected in the paper):
```
π = 0.01
 − log((1 − π)/π) = -4.59511985013459
```
- ResNet isn't pretrained. This seems very fixable. [My current solution](https://gist.github.com/c0nn3r/d0fcb7edb57c405cb707afeca28a8329) is to subclass ResNet and then override the `forward` method. This allows for the weights to be loaded, but for the network to have a completely different output. You structured your model different, so I'm not sure if this would work for your structure.
- COCO dataloader is missing. Seems like something that would be good to see if works, as VOC seems to work great (judging from loss).
- A detection script that can predict boxes on an image would be helpful to see the network's outputs. I cannot figure out how to do this in an elegant way, so would love your input on this. I'm confused how this implementation works with anchor points and how to transform the predictions back to image space.
- ~~Activation function should be Sigmoid~~ You fixed this :+1: 
- Learning rate training schedule does not match the paper, but the paper was working with a pretrained `ResNet` and training on the COCO dataset.
- The RetinaNet paper recommends flip dataset augmentation - seems very easy to add.

Anway, just let me know what you think and I can open a pull request for these things if you would like. I have written most of these fixes for work and would love to merge into your repo.

**Comment by Mendel1** [comment]
@c0nn3r  Have you tried to test the net for detection? Does it still work great ?I trained the net and get very small loss.But when I tried to predict the boxes and labels,I got all maximum scores for label 0(background).

**Comment by Mendel1** [comment]
In fact ,pretrain the model is easy for this net too.Just need to change the structure name(except top->bottom layers) the same as  original resnet.And you can load the dict you need from pretrained model .

**Comment by c0nn3r** [comment]
@Mendel1 Cool, I will take a look at that method. I still like the idea of breaking the ResNet and the feature pyramid apart, that way it will be easier in the future to switch to the different base network.

If I cannot see the output detection boxes, then I assume the network doesn't work. :wink: I think something is off calculating the loss, seems to be doing something fishly.

**Comment by Mendel1** [comment]
@c0nn3r  May I ask you a question that my sigmoid layer outputs zero somtimes？Do you know why that could happen?It should be between 0 and 1.:sob:

**Comment by c0nn3r** [comment]
If something is messed up with the loss function it could be optimal to output 0 each time. Also, are you training on the VOC dataset? Currently, I'm trying to get that to work first.

**Comment by Mendel1** [comment]
Yes,I am training on VOC with only 120 pictures.

**Comment by c0nn3r** [comment]
Doesn't it overfit if you run it for many epochs?

**Comment by Mendel1** [comment]
I want it to ovefit and therefore I can know if it can work .

**Comment by Mendel1** [comment]
I don't quite understand you.Could you explain it more specifically？Thanks a lot.

> If something is messed up with the loss function then it could be optimal to output zero each time.

**Comment by c0nn3r** [comment]
If the loss function doesn't work as it should, the network cannot minimize it and therefore leading to improper results. I'm going fork this repo and start trying to figure out what is going on. I really need to understand how the anchor box output works, however.

**Comment by Mendel1** [comment]
You can look at the decode function in encoder.py.  But I don't quite agree on that.

**Comment by c0nn3r** [comment]
Fair enough, I'm just guessing at this point.

**Comment by njtuzzy** [comment]
@mendel1 when you change the softmax to sigmoid, do you use the crossentropyloss instead of NLLoss ??

**Comment by Mendel1** [comment]
@njtuzzy Yes.I just replace the NLLoss with torch.sum( ).Is this wrong?

**Comment by Mendel1** [comment]
@c0nn3r  
To find how it crashed into 0.I changed the code below.
```python
        masked_cls_preds = cls_preds[mask].view(-1,self.num_classes)
        print('#'*9)
        print(masked_cls_preds[masked_cls_preds==0])
        print('#'*9)
        cls_loss = self.focal_loss(masked_cls_preds, cls_targets[pos_neg])
        
```
This is result.
```
Epoch: 0
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
58845.9453125
scratch :     868.00 | loc_loss:    1.354 | cls_loss:   67.795 | train_loss:   69.149 | avg_loss:   69.149
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
55631.6328125
scratch :     964.00 | loc_loss:    1.452 | cls_loss:   57.709 | train_loss:   59.161 | avg_loss:   64.155
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
49687.015625
scratch :    1096.00 | loc_loss:    1.377 | cls_loss:   45.335 | train_loss:   46.712 | avg_loss:   58.340
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
42116.0429688
scratch :    1195.00 | loc_loss:    1.558 | cls_loss:   35.244 | train_loss:   36.802 | avg_loss:   52.956
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
34168.3828125
scratch :    1059.00 | loc_loss:    1.524 | cls_loss:   32.265 | train_loss:   33.789 | avg_loss:   49.122
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
26225.1582031
scratch :    1022.00 | loc_loss:    1.433 | cls_loss:   25.661 | train_loss:   27.094 | avg_loss:   45.451
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
19361.9648438
scratch :     934.00 | loc_loss:    1.531 | cls_loss:   20.730 | train_loss:   22.261 | avg_loss:   42.138
#########
Variable containing:[torch.cuda.FloatTensor with no dimension]
#########
13987.390625
scratch :     994.00 | loc_loss:    1.362 | cls_loss:   14.072 | train_loss:   15.434 | avg_loss:   38.800

Test
#########
Variable containing:
 0
.....
 0
[torch.cuda.FloatTensor of size 441 (GPU 0)]

#########
10083.6582031
scratch :     995.00 | loc_loss:    1.080 | cls_loss:   10.134 | test_loss:   11.214 | avg_loss:   11.214
#########

```
It occurs first in test epoch.I don't know why.In test epoch,the net doesn't try to minimize it.It should just compute loss and print it.

**Comment by Mendel1** [comment]
Following my senior's advice,I add a small number to log function and thus the loss won't  be inf.But I still don't know why sigmoid can produce 0 for some situations.

**Comment by njtuzzy** [comment]
@Mendel1 Sigmoid is just use for binary class, so the out probilities is just the binary confidence for "a anchor" is/isnot "a class", am i right?

**Comment by Mendel1** [comment]
I think you're right.

**Comment by njtuzzy** [comment]
so my next question is when a anchor has two classes (for example) how to decide the anchor belong to which classes? just select the higher problility value?

**Comment by Mendel1** [comment]
I think if a anchor have overlaps more than the threshold with other anchor for specific class,the confidence for that class will be invalid.After this process,you can pick the class with higher probability.Thus my opinion  is do nms at first and then pick the class with higher probability for each anchor.But in kuangliu 's program,he select  the class at first and then do nms.So maybe you should review the paper and try to figure out the true way.

**Comment by njtuzzy** [comment]
will try it  thanks

**Comment by c0nn3r** [comment]
@Mendel1 Does the decode function work for you? Doesn't seem to work for me.

**Comment by Mendel1** [comment]
@c0nn3r No.Inmy situation,the decode function indicates for all box,confidence of background is the best and thus error occurs with positive label empty.

**Comment by andreaazzini** [comment]
@c0nn3r Have you managed to plug your own implementation of RetinaNet? I am currently working on fixing the encoder, which has a few issues.

**Comment by c0nn3r** [comment]
@andreaazzini I have a deadline coming up, so I don't have time to fix this repo. I am going to post all of my code for my own implementation soon.

**Comment by Mendel1** [comment]
The nms function in this project is wrong.NMS should work on box with the same class not only  overlaps.

**Comment by Mendel1** [comment]
Also,while predicting，the anchor should be reduced to 1K at most for each scale.This action should be performed before it's merged .

**Comment by njtuzzy** [comment]
@Mendel1 where is the predicting function? i dont find one?

**Comment by njtuzzy** [comment]
@kuangliu @Mendel1  In the encoder file, the output of "get_anchor_boxes" is the "xcenter,ycenter, xwidth, ywidth" format, it seem that it do not need to change to xxwh using change_box_order function?

**Comment by Mendel1** [comment]
You can write the predicting function  with decode function and the net .

**Comment by Mendel1** [comment]
@njtuzzy You are right.

**Comment by c0nn3r** [comment]
Turns out my deadline is a week later. I will see if this weekend I can publish my own implementation.

**Comment by kuangliu** [comment]
I'm closing this now. It's too long to read. 
Please move to #11 .

**Comment by njtuzzy** [comment]
@c0nn3r cheers:)
